src/Development/IDE/GHC/Warnings.hs

```diff
@@ -27,8 +28,17 @@ withWarnings diagSource action = do
   warnings <- newVar []
   let newAction :: DynFlags -> WarnReason -> Severity -> SrcSpan -> PprStyle -> SDoc -> IO ()
       newAction dynFlags wr _ loc style msg = do
-        let wr_d = fmap (wr,) $ diagFromErrMsg diagSource dynFlags $ mkWarnMsg dynFlags loc (queryQual style) msg
+        let wr_d = map ((wr,) . third3 (attachReason wr)) $ diagFromErrMsg diagSource dynFlags $ mkWarnMsg dynFlags loc (queryQual style) msg
         modifyVar_ warnings $ return . (wr_d:)
   res <- action $ \x -> x{ms_hspp_opts = (ms_hspp_opts x){log_action = newAction}}
   warns <- readVar warnings
   return (reverse $ concat warns, res)
+
+attachReason :: WarnReason -> Diagnostic -> Diagnostic
+attachReason wr d = d{_code = StringValue <$> showReason wr}
```

Comment:
This seems pretty weird, especially since the specification says this field may be displayed to the user. Perhaps the reason can be recorded in the ide state, and the code action can be generated based off of that?

Reply:
> This seems pretty weird, especially since the specification says this field may be displayed to the user.

Hmm, yeah, that is unfortunate. I was originally intending to use the more human-readable form here, analogously to how GHC shows e.g. `-Wunused-imports` in the CLI output. But this way it was easier to be sure that a particular diagnostic comes from a GHC warning (using `T.stripPrefix "Opt_Warn"`).

> Perhaps the reason can be recorded in the ide state, and the code action can be generated based off of that?

That sounds promising. There's a `diagnostics` field in `ShakeExtras` but the `DiagnosticStore` type doesn't seem to give us anywhere to store extra information, so I guess it would need replacing - I haven't yet looked at how much of a knock-on effect that might have.

Reply:
I'm struggling to see a nice way to put this information in `ShakeExtras`.

Besides, based on other LSP servers, I think this actually is more or less the intended use case for the `code` field. I'm thinking we put the full flag there e.g. `-Wunused-imports`, and trigger the action for diagnostics whose code begins with "-W" Still a little bit ugly, but the very nature of LSP means we're bound to be resigned to the occasional but of string-ly typed code.

The way VScode at least shows this is actually quite nice: `typecheck(-Wunused-imports)`. Regardless of the original purpose of this PR, it's nice to get this extra information about where a warning is coming from. In fact, it's the one thing in GHC's output that we're currently missing.

Reply:
> I'm thinking we put the full flag there e.g. -Wunused-imports

This is what c2cd237 does.

Reply:
The `code` field is intended to refer to the "handbook code" for a diagnostic, as provided by some compilers.  Version 3.16 of the spec has a `data` field for this, prompted from https://github.com/microsoft/language-server-protocol/issues/887.  Perhaps use that instead.

Reply:
I think seeing as the flag is the closest thing to a "handbook code" GHC is likely to have any time soon, this is still the right place to put it. Especially given that the result is visually appealing in VScode (haven't tried other editors, but I'd hope they show something relatively similar).

Obviously I'll change this if you or any of the other core maintainers feel strongly.